File: packages/guards/src/personal-opinion.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface PersonalOpinionOptions {
  action: 'block' | 'warn';
}

const OPINION_PATTERNS: RegExp[] = [
  /I (?:think|believe|feel|personally|prefer|suggest)\b/gi,
  /in my (?:opinion|view|experience)\b/gi,
  /I would (?:say|recommend|suggest)\b/gi,
  /my (?:favorite|preference|recommendation)\b/gi,
  /I(?:'m| am) (?:not a fan|a big fan)\b/gi,
  /from my perspective\b/gi,
  /if (?:you ask me|I were you)\b/gi,
];

export function personalOpinion(options: PersonalOpinionOptions): Guard {
  return {
    name: 'personal-opinion',
    version: '0.1.0',
    description: 'Detect personal opinions in factual responses',
    category: 'content',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of OPINION_PATTERNS) {
        const re = new RegExp(pattern.source, pattern.flags);
        const m = re.exec(text);
        if (m) matched.push(m[0]);
      }

      const triggered = matched.length > 0;

      return {
        guardName: 'personal-opinion',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered
          ? `Personal opinion detected: "${matched[0]}"`
          : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { matched, reason: 'Response contains personal opinion language instead of objective facts' }
          : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/personal_opinion.py
"""Detects personal opinion expressions."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_OPINION_PATTERNS = [
    r"\bi think\b",
    r"\bi believe\b",
    r"\bin my opinion\b",
    r"\bi feel\b",
    r"\bpersonally\b",
    r"\bmy view is\b",
]


class _PersonalOpinion:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "personal-opinion"
        self.action = action

    def check(self, text: str, stage: str = "output") -> GuardResult:
        start = time.perf_counter()
        lower = text.lower()
        found: List[str] = []
        for pattern in _OPINION_PATTERNS:
            matches = re.findall(pattern, lower)
            found.extend(matches)
        triggered = len(found) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name=self.name,
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Personal opinion detected: {', '.join(found)}" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matches": found, "count": len(found)} if triggered else None,
        )


def personal_opinion(*, action: str = "warn") -> _PersonalOpinion:
    return _PersonalOpinion(action=action)
